examples/llm/glem.py

```diff
@@ -0,0 +1,440 @@
+        model.lm = model.lm.to('cpu', non_blocking=True)
+    torch.cuda.empty_cache()
+
+    pretrain_phase_time = time.time() - pretrain_start_time
+    print(f'Pretrain {pretrain_phase} time: {pretrain_phase_time:.2f}s')
+    os.makedirs(osp.dirname(preds_dir), exist_ok=True)
+    torch.save(preds, osp.join(preds_dir, f'{preds_filename}.pt'))
+    print(
+        f'Saved predictions to {osp.join(preds_dir, f"{preds_filename}.pt")}')
+    train_acc, val_acc, test_acc = evaluate(preds, ['train', 'valid', 'test'])
+    print(f'Pretraining acc: {train_acc:.4f}, Val: {val_acc:.4f}, '
+          f'Test: {test_acc:.4f}')
+
+    # EM iterations
+
+    em_phase = em_order
+    """
+    We run E-step(LM training) and M-Step(GNN training) alternatively in each
+    em iterations, so the total number of iterations is num_em_iter * 2 and
+    we switch the em_phase at end of each iteration in following loop
+    """
+    for em_it in range(1, num_em_iters * 2 + 1):
+        pseudo_labels = preds.argmax(dim=-1)
+        best_val_acc = final_test_acc = 0.0
+        print(f'EM iteration: {em_it}, EM phase: {em_phase}')
+        optimizer = load_model(em_phase)
+        num_epochs = lm_epochs
+        train_loader = text_train_loader
+        test_loader = text_data_loader
+        early_stopping = 0
+        if em_phase == 'gnn':
+            train_loader = graph_train_loader
+            num_epochs = gnn_epochs
+            test_loader = subgraph_loader
+        for epoch in range(1, num_epochs + 1):
+            acc, loss = model.train(em_phase, train_loader, optimizer,
+                                    pseudo_labels, epoch, True, verbose)
+            if epoch >= 5 or epoch == num_epochs:
+                cur_preds = model.inference(em_phase, test_loader,
+                                            verbose=verbose)
+                train_acc, val_acc, test_acc = evaluate(
+                    cur_preds, ['train', 'valid', 'test'])
+
+                print(f'Train: {train_acc:.4f}, Val: {val_acc:.4f}, '
+                      f'Test: {test_acc:.4f}')
+
+                if val_acc <= best_val_acc:
+                    early_stopping += 1
+                    if early_stopping > patience:
+                        print(f'''Early stopped by Epoch: {epoch}, \
+                            Best acc: {final_test_acc}''')
+                        break
+                else:
+                    best_val_acc = val_acc
+                    final_test_acc = test_acc
+                    preds = cur_preds
+
+        if em_phase == 'gnn':
+            gnn_test_acc = max(gnn_test_acc, final_test_acc)
+            model.gnn = model.gnn.to('cpu', non_blocking=True)
+            em_phase = 'lm'
+        else:
+            lm_test_acc = max(lm_test_acc, final_test_acc)
+            model.lm = model.lm.to('cpu', non_blocking=True)
+            em_phase = 'gnn'
+        torch.cuda.empty_cache()
+    print(f'Best GNN acc: {gnn_test_acc}, LM acc: {lm_test_acc}')
```

Comment:
This is the same comment as https://github.com/pyg-team/pytorch_geometric/pull/9467#discussion_r1786532509, but we shouldn't pick the best metric evaluated on the test set at the end of every EM step.

Reply:
Hi Akihiro,

Thanks for reviewing the code.

I think the case is different here, I agree that we should not pick best test metrics after every epoch, but the test metrics still required for every EM step. Since E step is LM training and M-step is GNN training, both step need certain number of epochs. We need to run full inference after every E and M step for finding out that which model have better result.

Reply:
@ECMGit i think @akihironitta's point is that to "finding out that which model have better result" you should only use val accuracy, not test accuracy since if you use the test acc this could be viewed as a form of loosely fitting the model to the test set

Reply:
done